plugins/devlink/fu-devlink-device.c

```diff
@@ -0,0 +1,580 @@
+{
+	FuDevlinkDevice *self = FU_DEVLINK_DEVICE(device);
+
+	fwupd_codec_string_append(str, idt, "BusName", self->bus_name);
+	fwupd_codec_string_append(str, idt, "DevName", self->dev_name);
+}
+
+static gboolean
+fu_devlink_device_setup(FuDevice *device, GError **error)
+{
+	/* Get device information and version */
+	return fu_devlink_device_get_info(device, error);
+}
+
+/* Perform firmware activation using devlink reload with fw_activate action */
+static gboolean
+fu_devlink_device_activate(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuDevlinkDevice *self = FU_DEVLINK_DEVICE(device);
+	g_auto(FuDevlinkGenSocket) nlg = {};
+	struct nlmsghdr *nlh;
+
+	g_debug("activating firmware for %s/%s", self->bus_name, self->dev_name);
+
+	/* Open netlink socket */
+	if (!fu_devlink_netlink_gen_socket_open(&nlg, error))
+		return FALSE;
+
+	/* Prepare reload command with fw_activate action */
+	nlh = fu_devlink_netlink_cmd_prepare(&nlg, DEVLINK_CMD_RELOAD, FALSE);
+	mnl_attr_put_strz(nlh, DEVLINK_ATTR_BUS_NAME, self->bus_name);
+	mnl_attr_put_strz(nlh, DEVLINK_ATTR_DEV_NAME, self->dev_name);
+	mnl_attr_put_u8(nlh, DEVLINK_ATTR_RELOAD_ACTION, DEVLINK_RELOAD_ACTION_FW_ACTIVATE);
+
+	g_debug("sending devlink reload command with fw_activate action for %s/%s",
+		self->bus_name,
+		self->dev_name);
+
+	/* Send command and wait for response */
+	if (!fu_devlink_netlink_msg_send(&nlg, nlh, error)) {
+		g_prefix_error(error, "failed to send devlink reload command: ");
+		return FALSE;
+	}
+
+	g_debug("firmware activation completed for %s/%s", self->bus_name, self->dev_name);
+	return TRUE;
+}
+
+/**
+ * fu_devlink_device_new:
+ * @ctx: (nullable): optional #FuContext
+ * @bus_name: devlink bus name
+ * @dev_name: devlink device name
+ *
+ * Creates a new #FuDevlinkDevice.
+ *
+ * Returns: (transfer full): a #FuDevlinkDevice
+ */
+FuDevlinkDevice *
+fu_devlink_device_new(FuContext *ctx, const gchar *bus_name, const gchar *dev_name)
+{
+	g_autoptr(FuDevlinkDevice) self = NULL;
+	g_autofree gchar *device_id = NULL;
+
+	g_return_val_if_fail(bus_name, NULL);
+	g_return_val_if_fail(dev_name, NULL);
+
+	/* Create object and assign the strings */
+	self = g_object_new(FU_TYPE_DEVLINK_DEVICE, "context", ctx, NULL);
+	self->bus_name = g_strdup(bus_name);
+	self->dev_name = g_strdup(dev_name);
+
+	device_id = g_strdup_printf("%s/%s", bus_name, dev_name);
+	fu_device_set_physical_id(FU_DEVICE(self), device_id);
+	fu_device_set_name(FU_DEVICE(self), device_id);
+
+	return g_steal_pointer(&self);
+}
+
+static void
+fu_devlink_device_finalize(GObject *object)
+{
+	FuDevlinkDevice *self = FU_DEVLINK_DEVICE(object);
+
+	g_free(self->bus_name);
+	g_free(self->dev_name);
+
+	G_OBJECT_CLASS(fu_devlink_device_parent_class)->finalize(object);
+}
+
+static void
+fu_devlink_device_init(FuDevlinkDevice *self)
+{
+	fu_device_set_summary(FU_DEVICE(self), "Devlink device");
+	fu_device_add_protocol(FU_DEVICE(self), "org.kernel.devlink");
```

Comment:
You probably want `fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_MD_SET_FLAGS);` here

Reply:
Actually I don't. It's called in fu_devlink_component_init() I moved the flag definition to that file as well. The flag is relevant to component device only.